src/shell/mysh.c

```diff
@@ -9,13 +12,276 @@ typedef struct command {
 
 command * new_command() {
     command * cmd = (command *) malloc(sizeof(command));
-    cmd->cmd = NULL;
+    cmd->tokens = NULL;
     cmd->input = NULL;
     cmd->output = NULL;
     cmd->next = NULL;
     return cmd;
 }
 
-int main(int argc, char ** argv) {
+void print_cmd_list(command * cmd_pt) {
+    int i = 0;
+    int j;
+    for (; cmd_pt != NULL; cmd_pt = cmd_pt->next) {
+        printf("command %d:\n\
+  tokens:\n", i);
+        for (j = 0; cmd_pt->tokens[j] != NULL; j++) {
+            printf("    %s\n", cmd_pt->tokens[j]);
+        }
+        printf("  input: %s\n  output: %s\n", cmd_pt->input, cmd_pt->output);
+        i++;
+    }
+}
+
+/* Print a prompt containing the username and current working directory. */
+void print_prompt();
+
+/*
+ * Given a string from the command line, translate it into a linked list
+ * of cmd_struct's.
+ * First tokenize the string according to whitespace and double quotes,
+ * then structurize the tokens,
+ * structures demarked by pipes ("|") and redirects ("<", ">").
+ */
+command * structure_cmds(char *);
+
+/*
+ * Given an array of string pointers, a source string, and the number of tokens
+ * to make from the string,
+ * create new strings to represent each token and set the array elements
+ * to point to them. 
+ * Recognize newlines and string termination characters as end of string.
+ * If too many or too few tokens are found, report the error to stderr and exit.
+ */
+void split_cmd(char ** dst, char * src, char * end, int n_tokens);
+
+int main(void) {
+    /* maximum bytes in an input line */
+    int max_len = 1024;
+    char cmd_str[max_len];
+
+    command * cmd_list;
+    while(1) {
+        print_prompt();
+
+        fgets(cmd_str, max_len, stdin);
+
+        cmd_list = structure_cmds(cmd_str);
+        print_cmd_list(cmd_list);
+    }
+
+
     return 0;
 }
+
+void print_prompt() {
+    char cwd[1024];
+    getcwd(cwd, sizeof(cwd));
+    printf("%s:%s>", getlogin(), cwd);
+}
+
+command * structure_cmds(char * cmd_str) {
+    char *begin = cmd_str;
+    char *iter = cmd_str;
+    /* possible states are adding a command (0), 
+     * redirecting input (1), and redirecting output (2)
+     */
+    char state = 0;
```

Comment:
Good place to use an enum: http://stackoverflow.com/questions/1102542/how-to-define-an-enumerated-type-enum-in-c